### PR TITLE
Small refactor to marshal code to support upcoming "set" tag

### PR DESCRIPTION
**Pull request by kalman**

### CODE REVIEW

**Comment by kalman** [comment]
I'm not going to finish this by the EOW so sending you a quick change. It'll make the real change easier to review.

File: go/marshal/decode.go [hunk 1]
```diff
@@ -12,21 +12,27 @@ import (
 	"github.com/attic-labs/noms/go/types"
 )
 
-// Unmarshal converts a Noms value into a Go value. It decodes v and stores the result
-// in the value pointed to by out.
+// Unmarshal converts a Noms value into a Go value. It decodes v and stores the
+// result in the value pointed to by out.
```
Comment (kalman) [inline go/marshal/decode.go]: I have this memory of @cmasone-attic or @rafael-atticlabs saying we were wrapping comments at 80 now.

For sure, without wrapping, the godoc (command line) tool isn't happy.

File: go/marshal/encode.go [hunk 2]
```diff
@@ -116,6 +131,12 @@ func (e *InvalidTagError) Error() string {
 	return e.message
 }
 
+type nomsTags struct {
+	name      string
+	omitEmpty bool
+	skip      bool
```
Comment (kalman) [inline go/marshal/encode.go]: `set` goes here

File: go/marshal/encode.go [hunk 3]
```diff
@@ -271,25 +292,30 @@ func (c *encoderCacheT) set(t reflect.Type, e encoderFunc) {
...
 	}
 
-	if !types.IsValidStructFieldName(name) {
-		panic(&InvalidTagError{"Invalid struct field name: " + name})
+	if !types.IsValidStructFieldName(tags.name) {
+		panic(&InvalidTagError{"Invalid struct field name: " + tags.name})
 	}
 
-	if idx != -1 {
+	if len(tagsSlice) > 1 {
```
Comment (kalman) [inline go/marshal/encode.go]: and here

File: go/marshal/encode.go [hunk 4]
```diff
@@ -18,27 +20,33 @@ import (
...
+// Floating point and integer values are encoded as Noms types.Number. At the
+// moment this might lead to some loss in precision because types.Number
+// currently takes a float64.
 //
 // String values are encoded as Noms types.String.
 //
 // Slices and arrays are encoded as Noms types.List.
 //
-// Maps are encoded as Noms types.Map
+// Maps are encoded as Noms types.Map, or if the value is struct{}, a types.Set.
```
Comment (arv) [inline go/marshal/encode.go]: but only with a set tag

Comment (kalman) [inline go/marshal/encode.go]: Oops yeah, comment left over from the half finished patch which implemented this anyway.

**Comment by kalman** [comment]
Btw, here is the comment thing I mentioned: https://github.com/golang/go/wiki/CodeReviewComments#line-length

**Pull request merged**